### ISSUE TITLE
Defining two or more SR-IOV interfaces that use the same underling resource is buggy

### ISSUE THREAD

**Issue by ormergi** [kubevirt/kubevirt#7444]
**What happened**:
A VM with two or more SR-IOV interfaces **from the same resource** pool might end up with each interface connected to the wrong network.

For example https://github.com/kubevirt/kubevirt/issues/6351#issuecomment-918467186

The problem seems to be the fact that there is no mapping between the SR-IOV device PCI address and the 
NetworkAttachmentDefinition, but only between the SR-IOV device PCI address to the resource-name. 
Therefore there is no differentiation between the devices and the mapping is inconsistent.

**What you expected to happen**:
Each SR-IOV interface should be attached to the correct network as specified in the VM manifest.

**How to reproduce it (as minimally and precisely as possible)**:

**Additional context**:
We discover this while investigating https://github.com/kubevirt/kubevirt/issues/6351

**Environment**:
- KubeVirt version (use `virtctl version`): N/A
- Kubernetes version (use `kubectl version`): N/A
- VM or VMI specifications: N/A
- Cloud provider or hardware configuration: N/A
- OS (e.g. from /etc/os-release): N/A
- Kernel (e.g. `uname -a`): N/A
- Install tools: N/A
- Others: N/A

**Comment by YitzyD** [comment]
We are experiencing this issue as well. 
Our setup:
- [multus](https://github.com/k8snetworkplumbingwg/multus-cni)
- [sriov-cni](https://github.com/k8snetworkplumbingwg/sriov-cni)

The issue seems to be related to https://github.com/k8snetworkplumbingwg/multus-cni/issues/499. As the order of host PCI addresses in the  environment variable `PCIDEVICE_XXX=` that is set by either multus or the sriov-cni (I have yet to investigate which controls the injected environment variable) is indeterminate, the virt-launcher converter cannot (and likely should not) depend on it for assigning hostdev source addresses to interface devices.

The issue is in in the following peices of code: 
1) 
https://github.com/kubevirt/kubevirt/blob/542fe4bf82a9eff32866452de17796fbd5a2b64c/pkg/virt-launcher/virtwrap/device/hostdevice/generic/addresspool.go#L34
leads to: 
https://github.com/kubevirt/kubevirt/blob/542fe4bf82a9eff32866452de17796fbd5a2b64c/pkg/virt-launcher/virtwrap/device/hostdevice/addresspool.go#L45 
- wherein virtwrap sets up a pool of PCI addresses from the `PCIDEVICE_XXX` environment variable to be used when generating the domain xml

&nbsp; 
2) 
https://github.com/kubevirt/kubevirt/blob/542fe4bf82a9eff32866452de17796fbd5a2b64c/pkg/virt-launcher/virtwrap/device/hostdevice/hostdev.go#L58
leads to: 
https://github.com/kubevirt/kubevirt/blob/542fe4bf82a9eff32866452de17796fbd5a2b64c/pkg/virt-launcher/virtwrap/device/hostdevice/hostdev.go#L88
- According to the order of interfaces in the VMI yaml, a PCI address from the pool is assigned to the hostdev source address

&nbsp; 
Example:

virt-launcher env:
```bash
PCIDEVICE_VF=0000:e1:0a.3,0000:e1:0b.0
KUBEVIRT_RESOURCE_NAME_sriov0=vf
KUBEVIRT_RESOURCE_NAME_sriov0=vf
```
vmi spec:
```yaml
...
interfaces:
  - name: sriov0
    sriov: {}
  - name: sriov1
    sriov: {}
```
sriov0 will be assigned to host PCI device: `0000:e1:0a.3`
sriov1 will be assigned to host PCI device: `0000:e1:0b.0`

This is irrespective of whether the VF programmed by the sriov cni for the PCI device is associated with the network attachment definition name in the vmi networks array.

---
### Solutions:
1) Patch the CNIs so that the order of PCI addresses assigned to the `PCIDEVICE_XXX` environment variable is in some determinable order.
2) Virt-launcher will have to use a different method of assigning the hostdev source PCI address so that it matches the network attachment definition name

**Comment by ormergi** [comment]
@YitzyD Thanks for sharing and detailed info :slightly_smiling_face:
To workaround this, you can create a VFs device pool (sriov-device-plugin config or `SriovNetwotkPolicy` if SR-IOV network operator deployed) for each group of VF's by their configuration [[1]](https://github.com/kubevirt/kubevirt/issues/6351#issuecomment-1077937853).
For example: one VFs pool that will be configured with VLAN 20 and another for VFs with VLAN 30.

In any case we plan to fix it properly.

**Comment by ormergi** [comment]
@YitzyD 
Fixed by https://github.com/kubevirt/kubevirt/pull/8226, available on versions: v0.53.0, v0.54.0, v0.55.0, v0.56.0 and above.
